pooch/downloaders.py

```diff
@@ -308,3 +314,113 @@ def callback(data):
+        requires authentication (i.e., no anonymous SFTP). Use the empty
+        string to indicate no password is required.
+    timeout : int
+        Timeout in seconds for sftp socket operations, use None to mean no
+        timeout.
+    progressbar : bool
+        If True, will print a progress bar of the download to standard
+        error (stderr). Requires `tqdm <https://github.com/tqdm/tqdm>`__ to
+        be installed.
+
+    """
+
+    def __init__(
+        self,
+        port=22,
+        username="anonymous",
+        password="",
+        account="",
+        timeout=None,
+        progressbar=False,
+    ):
+        if paramiko is None:
+            raise ImportError("SFTP is only supported if paramiko is installed")
+
+        self.port = port
+        self.username = username
+        self.password = password
+        self.account = account
+        self.timeout = timeout
+        self.progressbar = progressbar
+        if self.progressbar and tqdm is None:
+            raise ValueError("Missing package 'tqdm' required for progress bars.")
+
+    def __call__(self, url, output_file, pooch):
+        """
+        Download the given URL over SFTP to the given output file.
+
+        Parameters
+        ----------
+        url : str
+            The URL to the file you want to download.
+        output_file : str or file-like object
+            Path (and file name) to which the file will be downloaded.
+        pooch : :class:`~pooch.Pooch`
+            The instance of :class:`~pooch.Pooch` that is calling this
+            method.
+        """
+
+        parsed_url = parse_url(url)
+        isfile = hasattr(output_file, "write")
+
+        connection = paramiko.Transport(sock=(parsed_url["netloc"], self.port))
+
+        if isfile:
+            # paramiko requires a file name as a string, not a file object
+            # file is first closed to avoid any errors and then get name
+            output_file.close()
+            output_file = output_file.name
+        sftp = None
+        try:
+            connection.connect(username=self.username, password=self.password)
+            sftp = paramiko.SFTPClient.from_transport(connection)
+            sftp.get_channel().settimeout = self.timeout
+
+            if self.progressbar:
+                size = int(sftp.stat(parsed_url["path"]).st_size)
+                use_ascii = bool(sys.platform == "win32")
+                progress = tqdm(
+                    total=size,
+                    ncols=79,
+                    ascii=use_ascii,
+                    unit="B",
+                    unit_scale=True,
+                    leave=True,
+                )
+                with progress:
+
+                    def callback(current, total):
+                        "Update the progress bar and write to output"
+                        progress.total = int(total)
+                        progress.update(int(current - progress.n))
+
```

Comment:
As a side note: these lines of code appear to be common across the `HTTPDownloader`, `FTPDownloader`. It would be nice to do some refactoring at some point in the future so as to avoid repeating ourselves.

Reply:
Agreed. This is a good target for refactoring.